sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -205,14 +205,17 @@ object PhysicalAggregation {
     case logical.Aggregate(groupingExpressions, resultExpressions, child) =>
       // A single aggregate expression might appear multiple times in resultExpressions.
       // In order to avoid evaluating an individual aggregate function multiple times, we'll
-      // build a set of the distinct aggregate expressions and build a function which can
+      // build a map of the distinct aggregate expressions and build a function which can
       // be used to re-write expressions so that they reference the single copy of the
-      // aggregate function which actually gets computed.
-      val aggregateExpressions = resultExpressions.flatMap { expr =>
+      // aggregate function which actually gets computed. Note that aggregate expressions
+      // should always be deterministic, so we can use its canonicalized expression as its
```

Comment:
can you point out where the code is that guarantee this? I took a look at `CheckAnalysis` and seems Spark allow Aggregate to have non-deterministic expressions.

Reply:
@cloud-fan Agreed. e.g. `first()` in Spark SQL is marked as nondeterministic right now (although for the case of `first()` I'd actually believe we should make it deterministic instead, but that's for another story)

Reply:
So we are talking about two types of "non-deterministic" here:
1. Across-query non-deterministic but in-query deterministic, which means the same expression can produce different results over the same input between different runs, but should always give the same result within the same run. sum/avg on floating point numbers could be an example. Shall we make sure that "select sum(f) - sum(f) from t" always return 0? and similarly for "first()" maybe, should "select first_value(c) = first_value(c) over ..." always return true?
It is important to define the behavior first, which will lead to opposite approaches on how to handle the "deterministic" field here.
2. Across-query and in-query non-deterministic, which I don't think is allowed anyway.

Reply:
good point, I think `first` is a nondeterministic function, SQL Server also document it: https://docs.microsoft.com/en-us/sql/relational-databases/user-defined-functions/deterministic-and-nondeterministic-functions (search for FIRST_VALUE).

But here, although `first` is nondeterministic, I think `select first(a) - first(a) from ...` should return 0, as its result should be consistent in-query.

CC @gatorsmile @hvanhovell

Reply:
Then I think the solution we have right now (in my latest ci) is what we need. But we need to refine the code comment there to further illustrate cases like "first_value". And in general we may need better documentation regarding the behavior of FIRST_VALUE and similar functions. Agree?

Reply:
If we have the same non-deterministic functions in the same projection, they should return the same value.

Reply:
That's not true, see Postgres:
```
cloud=# select random(), random();
      random       |      random      
-------------------+------------------
 0.360924747306854 | 0.90011026058346
(1 row)

cloud=# select random() - random();
      ?column?       
---------------------
 -0.0260558221489191
(1 row)
```

Seems we need a way to distinguish FIRST_VALUE and RANDOM. Both of them are nondeterministic, but `FIRST_VALUE` has consistent values in the same project/aggregate.

Reply:
I did the check in Oracle. They return the same values.